OsmAnd-java/src/main/java/net/osmand/router/BinaryRoutePlanner.java

```diff
@@ -397,128 +400,125 @@ public static void printDebugMemoryInformation(RoutingContext ctx) {
-			segmentDist += squareRootDist(x, y, prevx, prevy);
-
-			// 2.1 calculate possible obstacle plus time
+		RouteSegment nextCurrentSegment = startSegment;
+		RouteSegment currentSegment = null;
+		while (nextCurrentSegment != null) {
+			currentSegment = nextCurrentSegment;
+			nextCurrentSegment = null;
 			
-			double obstacle = ctx.getRouter().defineRoutingObstacle(road, segmentPoint, (dir && !reverseWaySearch));
-			if (obstacle < 0) {
-				directionAllowed = false;
-				break;
-			}
-			double heightObstacle = ctx.getRouter().defineHeightObstacle(road, !reverseWaySearch ? prevSegmentPoint : segmentPoint, 
-					!reverseWaySearch ? segmentPoint : prevSegmentPoint);
-			if(heightObstacle < 0) {
+			// 1. calculate obstacle for passing this segment 
+			float segmentAndObstaclesTime = (float) calculateRouteSegmentTime(ctx, reverseWaySearch, currentSegment);
+			if (segmentAndObstaclesTime < 0) {
 				directionAllowed = false;
 				break;
 			}
-			boolean alreadyVisited = checkIfOppositeSegmentWasVisited(ctx, reverseWaySearch, graphSegments, segment, oppositeSegments,
-					prevSegmentPoint, segmentPoint, segmentDist, obstaclesTime);
-			obstaclesTime += obstacle;
-			obstaclesTime += heightObstacle;
-			if (alreadyVisited) {
-				directionAllowed = false;
-				break;
-			}
-			// correct way of handling precalculatedRouteDirection 
-			if (ctx.precalculatedRouteDirection != null) {
-//				long nt = System.nanoTime();
-//				float devDistance = ctx.precalculatedRouteDirection.getDeviationDistance(x, y);
-//				// 1. linear method
-//				 segmentDist = segmentDist * (1 + devDistance / ctx.config.DEVIATION_RADIUS);
-//				// 2. exponential method
-//				segmentDist = segmentDist * (float) Math.pow(1.5, devDistance / 500);
-				// 3. next by method
-//				segmentDist = devDistance ;
-//				ctx.timeNanoToCalcDeviation += (System.nanoTime() - nt);
-			}
-			// could be expensive calculation
-			// 3. get intersected ways
-			final RouteSegment roadNext = ctx.loadRouteSegment(x, y, ctx.config.memoryLimitation - ctx.memoryOverhead);
-			float distStartObstacles = segment.distanceFromStart + calculateTimeWithObstacles(ctx, road, segmentDist, obstaclesTime);
+			// calculate new start segment time as we're going to assign to put to visited segments
+			float distFromStartPlusSegmentTime = currentSegment.distanceFromStart + segmentAndObstaclesTime;
 			if (ctx.precalculatedRouteDirection != null && ctx.precalculatedRouteDirection.isFollowNext()) {
-				// reset to f
-//				distStartObstacles = 0;
-				// more precise but slower
-				distStartObstacles = ctx.precalculatedRouteDirection.getDeviationDistance(x, y) / ctx.getRouter().getMaxSpeed();
+				// speed up calculation with calculated route by using different distance from start
+				final int x = road.getPoint31XTile(currentSegment.getSegmentStart());
+				final int y = road.getPoint31YTile(currentSegment.getSegmentStart());
+				// TODO double check is it correct (looks OK)
```

Comment:
TODO

Reply:
It's never use, because ctx.precalculatedRouteDirection.isFollowNext() always false

Reply:
TODO delete

Reply:
**old c++**
routing_time="20498.07" loadedTiles="28399" visitedSegments="1176679" complete_distance="615890.50" complete_time="22612.58" 
**new c++**
routing_time="20502.49" loadedTiles="34978" visitedSegments="1519501" complete_distance="615886.44" complete_time="22612.42"